R/utils.R

```diff
@@ -380,3 +380,96 @@ defunction <- function(x) {
+#'
+#' @param lockfile lock file name passed to `pak::lockfile_create()`
+#'
+#' @return (`character(n)`) content of `lockfile`.
+#'
+#' @keywords internal
+create_lockfile_future <- function(lockfile = "session.lock") {
+  future::future(
+    {
+      packages <- sessioninfo::session_info()$packages
+      # # FOR TESTING return(capture.output(print(packages))) # FOR TESTING
+
+      # # Check '?pkgdepends::pkg_refs()'
+      # to see why we need to append custom syntax in package names,
+      # if they do not come from CRAN-like repository.
+
+      cran_like <- packages$package[grepl("CRAN|Bioconductor", packages$source)]
+
+      local <- packages[grepl("local", packages$source), ]
+      local <- apply(local, MARGIN = 1, function(row) {
+        sprintf(
+          "%s=local::%s",
+          row[1], # pacakge,
+          row[5] # loadedpath
+        )
+      })
+
+      github <- packages[grepl("Github", packages$source), ]
+      gitlab <- packages[grepl("Gitlab", packages$source), ]
+      github$source <- gsub("\\)$", "", gsub("Github (", "", grep("Github", github$source, value = TRUE), fixed = TRUE))
+      gitlab$source <- gsub("\\)$", "", gsub("Gitlab (", "", grep("Gitlab", gitlab$source, value = TRUE), fixed = TRUE))
+      github <- apply(github, MARGIN = 1, function(row) {
+        sprintf(
+          "%s=%s",
+          row[1], # pacakge,
+          row[9] # source
+        )
+      })
+      gitlab <- apply(gitlab, MARGIN = 1, function(row) {
+        sprintf(
+          "%s=%s",
+          row[1], # pacakge,
+          row[9] # source
+        )
+      })
+
+      # Custom installed packages are needed for the case where a basic teal app is started,
+      # and there are no teal.reporter calls, so no teal.reporter is not loaded.
+      # But teal.reporter is needed for local teal installations, and it needs to be checked
+      # if this was installed from a custom repository.
+      custom_installed <- sessioninfo::session_info("installed")$packages
+      custom_installed <- custom_installed[grepl("http", custom_installed$source), ]
+      custom_installed$source <- gsub("\\s.*", "", custom_installed$source) # remove R version
+      if (nrow(custom_installed) > 0) {
+        old_repos <- options("repos")
+        on.exit(options(repos = old_repos$repos))
+
+        # Extend options(repos) with a custom repository for pak::lockfile_create
+        options(
+          repos = unique(c(
+            options("repos")$repos,
+            custom_installed$source
+          ))
+        )
+      }
+
+      custom <- packages[grepl("http", packages$source), ]
+      custom$source <- gsub("\\s.*", "", custom$source) # remove R version
+      custom <- apply(custom, MARGIN = 1, function(row) {
+        sprintf(
+          "%s=url::%s/src/contrib/%s_%s.tar.gz",
+          row[1], # pacakge
+          row[9], # source
+          row[1], # pacakge
+          row[3] # loadedVersion
+        )
+      })
```

Comment:
Why do we have to do this all? Can't we just create a lock file _as is_ without any modifications on the top of it? 

All of these modifications introduces quite a lot of points of failures - e.g. I can see that you cover GitHub and GitLab - how about BitBucket, git, url, Bioc etc? I think I can go easily and challenge this more.

I just don't want to take responsibility of this all as this is quite complex and I think that it's quite likely that `pak` does it for us.

Reply:
Yeah @pawelru so that is what you misunderstood - `pak` does not do that for us. It only does it for you if you have a `package-tree structure`. If you want to create a lockfile of the session then you need to specify packages on your own, the way I did in this function.

By default it assumes it takes package directory

<img width="695" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/fc6a4b1e-782e-436b-82c6-9733a999978f">

And if you are in the shiny app or a regular R project directory if fails on a default run because of that

<img width="636" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/11845373-fba2-47cd-ade5-4c21705fa4ac">

Reply:
@pawelru if you don't want us to handle all the responsibility they we should try `renv::snapshot()` that actually does a lot for us.

Reply:
> how about BitBucket, git, url, Bioc etc? I think I can go easily and challenge this more.

Yeah but you could also easily read the comment I left about `?pkgdepends::pkg_refs()`

> # # Check '?pkgdepends::pkg_refs()'
> # to see why we need to append custom syntax in package names,
> # if they do not come from CRAN-like repository.
 
 This is what is currently supported
 
 
<img width="455" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/1d5fbd8a-b1e1-49f2-b5ba-8a8b03b6ffb4">


Bitbucket doesn't have any custom treatment so would be treated as `local` or `url`.
`url` is covered by `custom` in my code
`git` will be treated the same way as `custom` - you will just get the path to the git folder.

# Anyway

I don't like it neither. And I support @vedhav concern https://github.com/insightsengineering/teal/pull/1224#issuecomment-2125005923

> And we'd have to test this solution for different deployment types as well

Imagine how many different package source installation configuration we need to test to provide this functionality on our end. If we just let the lockfile content to be passed by the app developer, then it removes a huge burden from our shoulder. I think it's easier for the app developer to learn on how to create a lockfile, then on us to cover all cases with packages being installed locally, from custom repositories, various versions of Bioconductor, CRAN, MRAN SNAPSHOTS, from Bitbucket, from Gitlab/Github, from AWS S3 and Microsoft Azure Blob Storage where you often need to have an authentication to access the url.

> I think I can go easily and challenge this more.

 : P I think it was naive to assume `pak::lockfile_create()` magically does everything for us and we are covered

Reply:
OK I see it now. I'm seeing two ways now:
- take packages from sessioninfo, create a dummy DESCRIPTION file and put them in the Imports and use `pak` for creating lock file
- use `renv` from the very beggining for dependency discovery as well as lock file creation

Reply:
Will try both - thanks for the guidance

Reply:
> * take packages from sessioninfo, create a dummy DESCRIPTION file and put them in the Imports and use `pak` for creating lock file

I think it might be more robust in a long term - instead of handling these sources ourselves (through `sessioninfo` package)

Reply:
Yeah, but you will need to figure out how to include sources in a DESCRIPTION file : ) the most problematic sources to be included in DESCRIPTION
- source for custom repository 
- for a package installed locally 
- AWS S3 buckets / Azure Storage Blobs
- Bitbucket
- Azure DevOps repositories (like GitHub / GitLab / Bitbucket)


I think we will have a mixture of DESCRIPTION file and custom editions to `options(repos)` setting. Just trying to gameplan this - I don't have yet the final form in my head

Reply:
> the most problematic sources to be included in DESCRIPTION
> ...
> - for a package installed locally

I guess this could be considered as a violation of a reproducibility

> Yeah, but you will need to figure out how to include sources in a DESCRIPTION file. 

I see the point and I think this is a real problem too. Normally, DESCRIPTION should contain a `Repository` field with name of the repo from current `options("repos")`. I consider deploying an application with package from a repo not available in `options("repos")`  as a violation of reproducibility. However we can still help app-developers with `sessioninfo` - it seems to address this issue and is able to figure-out location somehow.

Reply:
Hey guys, I decided to prepare a solution with `renv::snapshot()` as I found recreation of the `sessioninfo::session_info()` into a DESCRIPTION file that can be used by `pak::lockfile_create` challenging due to possible multiple custom sources. `pak::lockfile_create` fails if it can not install even one package, where `renv::snapshot()` does not try to install packages and for the problematic ones it just dumps the version of the package into the `.lockfile` with an unknown source.